fix: [node npm] service has bad colors #4809

regevbr opened the pull request:
Fix #4809

regevbr commented:
Instead of getting the current version (which can be an odd version which is unstable and will never be LTS), I get all the current LTS versions and make sure that the package matches to all of them

regevbr commented:
The dockercloud tests failed for some reason and the link is broken so I can't check why...

calebcartwright commented:
> The dockercloud tests failed for some reason

Don't worry about that one, it's unrelated #4808

calebcartwright commented:
The failing service test is due to changes in this PR though. Can you please take a look whenever you get a chance @regevbr 

https://circleci.com/gh/badges/shields/94808?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

```bash
NodeVersion [live] gets the node version of passport [ GET /passport.json ] - [ GET /passport.json ]
  AssertionError: expected [Function] to not throw an error but 'TypeError: Invalid comparator: intermediate' was thrown
```

regevbr commented:
@calebcartwright this is actually doesn't seem related to me - the message the live tests return is "credentials are misconfigured" and it is also mentioned in the link you sent me.
I'm trying to run it locally and it fails as well, but I'm not sure what those credentials should be... Can you please help?

regevbr commented:
@calebcartwright I managed to find the reason for this failure, and it is not due to my changes - one or both of NPM_ORIGINS and NPM_TOKEN env variables are missing in the ci process...
After setting those in my dev env, the tests are passing.

calebcartwright commented:
Yeah, seems like the node failure was spurious. Thanks for checking!

Will take a closer look at this PR over the weekend

regevbr commented:
@calebcartwright I have implemented a new badge.
Tested in locally and it works great.
Can you please CR?

regevbr commented:
seems like tests are failing due to issues with github integration and not related to my code

calebcartwright commented on services/node/node-lts.service.js:
```diff
@@ -0,0 +1,18 @@
+  static get path() {
+    return 'v-lts'
+  }
+
```
Should probably have different default badge labels between these two services, `node lts` for this one for example, so `defaultBadgeData` may be better off specified explicitly in each respective badge class

regevbr replied:
Will do

calebcartwright commented on services/node/node-version-color.spec.js:
```diff
@@ -0,0 +1,54 @@
+describe('node version color', function() {
+  describe(`stable`, function() {
+    it('should only return lts versions', async function() {
+      this.timeout(4000)
```
I'd really like to avoid running tests like this that are making live API calls within our unit test suite. It'd probably better to test the logic either by setting up [service tests using nock to mock the API responses](https://github.com/badges/shields/blob/master/doc/service-tests.md#5-mocking-responses) or perhaps keeping these as unit tests but using Sinon to mock out the relevant function calls

regevbr replied:
Will do

calebcartwright commented:
Thanks for the quick response! Will review later today

regevbr commented:
@calebcartwright thanks,
 it is not ready yet though, will let you know when it is. It is a lot of work to get all the mockings done as there are many mutations to it.

You should know I found a bug in the npm verison manager where when asked to get version of package with a certain tag (and not scoped) it actually got the latest version info.
I fixed the bug in this PR as well.

regevbr commented:
@calebcartwright I'm done, waiting for your CR

calebcartwright commented:
> You should know I found a bug in the npm verison manager where when asked to get version of package with a certain tag (and not scoped) it actually got the latest version info.
I fixed the bug in this PR as well.

Is it possible to open a separate PR with that change, or do you need those changes here to update the node platform support badges?

regevbr commented:
Sadly I can't separate...

calebcartwright commented:
> Sadly I can't separate...

Ah sorry, forgot the node classes extended the npm base class

calebcartwright commented:
> You should know I found a bug in the npm verison manager where when asked to get version of package with a certain tag (and not scoped) it actually got the latest version info.
I fixed the bug in this PR as well

Would you elaborate a bit more on what you believe to be a bug. Is it just for the node service and/or with any npm badges? As noted in the inline comment, the call with `/latest` in `fetchPackageData` was done deliberately for those npm badges to minimize the response size and because the npm badges that consume that function (collaborators, license, dependency version, etc.) don't need or support the `tag` filtering.

The npm version badge is working correctly, but it's not using that same function:

https://github.com/badges/shields/blob/5de4334ea709f9e8f0b464b18c03009bdc3f4b50/services/npm/npm-version.service.js#L92-L96

https://img.shields.io/npm/v/express/next
![](https://img.shields.io/npm/v/express/next)
https://img.shields.io/npm/v/express/latest
![](https://img.shields.io/npm/v/express/latest)

If the bug is only for the node platform support badges, then we should find a different way to fix that which doesn't impact all the other npm badges unnecessarily.

calebcartwright commented on services/node/testUtils/packageJsonVersionsTemplate.json:
```diff
@@ -0,0 +1,662 @@
+{
```
Could we trim this down to the minimal set of keys that are required? It'd really increase readability to only have the core pieces in here that are needed for the tests

regevbr replied:
sure thing

calebcartwright commented on services/node/testUtils/test-utils.js:
```diff
@@ -0,0 +1,184 @@
+
+const getTemplate = template => JSON.parse(templates[template])
+
+const mockPackageData = (
+  packageName,
+  engines,
+  scope,
+  tag,
+  registry
+) => nock => {
```
Let's destructure here given the quantity of params so that we don't have to rely on position and passing a lot of falsy values unnecessarily when there's a need to provide a value for `tag` or `registry`
```suggestion
const mockPackageData = ({
  packageName,
  engines,
  scope,
  tag,
  registry,
}) => nock => {
```

regevbr replied:
yes my bad, will fix it

calebcartwright commented on services/node/node-current.tester.js:
```diff
@@ -0,0 +1,112 @@
+function expectSemverRange(message) {
+  expect(() => new Range(message)).not.to.throw()
+}
+
```
Let's make sure we do still have live tests as well to provide that level of integration validation our service tests usually do. The mocked tests are great for this uncommon case where it's pretty tricky to test some of the internal logic within `node-version-color` with unit tests, but we definitely want to have the live tests as well.

regevbr replied:
sure thing

calebcartwright commented on services/node/node-lts.tester.js:
```diff
@@ -0,0 +1,166 @@
+function expectSemverRange(message) {
+  expect(() => new Range(message)).not.to.throw()
+}
+
```
Same comment as in `node-current.tester.js` regarding live tests

regevbr replied:
sure thing

calebcartwright commented on services/node/node-lts.tester.js:
```diff
@@ -0,0 +1,166 @@
+  .intercept(mockPackageData(`passport`, `10 - 12`))
+  .intercept(mockVersionsSha())
+  .expectBadge({ label: 'node lts', message: `10 - 12`, color: `brightgreen` })
+  .afterJSON(json => {
+    expectSemverRange(json.message)
+  })
```
Do we need this for these mocked tests given that `expectBadge` is already doing an exact match on the expected message?

regevbr replied:
Nope, I will remove them in mocked tests

regevbr commented:
@calebcartwright I fixed all your remarks.

Regarding your question:

> > You should know I found a bug in the npm verison manager where when asked to get version of package with a certain tag (and not scoped) it actually got the latest version info.
> > I fixed the bug in this PR as well
> 
> Would you elaborate a bit more on what you believe to be a bug. Is it just for the node service and/or with any npm badges? As noted in the inline comment, the call with `/latest` in `fetchPackageData` was done deliberately for those npm badges to minimize the response size and because the npm badges that consume that function (collaborators, license, dependency version, etc.) don't need or support the `tag` filtering.
> 
> The npm version badge is working correctly, but it's not using that same function:
> 
> https://github.com/badges/shields/blob/5de4334ea709f9e8f0b464b18c03009bdc3f4b50/services/npm/npm-version.service.js#L92-L96
> 
> https://img.shields.io/npm/v/express/next
> ![](https://camo.githubusercontent.com/429e5753fc696ca6d068c9b9aca46127db713b9e/68747470733a2f2f696d672e736869656c64732e696f2f6e706d2f762f657870726573732f6e657874)
> https://img.shields.io/npm/v/express/latest
> ![](https://camo.githubusercontent.com/3d568fe1c8d8c7c5f161def292340189a978205e/68747470733a2f2f696d672e736869656c64732e696f2f6e706d2f762f657870726573732f6c6174657374)
> 
> If the bug is only for the node platform support badges, then we should find a different way to fix that which doesn't impact all the other npm badges unnecessarily.

In order to get npm package information on a specific version or tag, one needs to use the `${registryUrl}/${packageName}` endpoint, which returns ALL the versions of the package and their info. In it, there is a list of tags (`dist-tags`) which you can use to resolve actual version of the latest release, or a custom tag and then get the info on it from the `versions` key.
This endpoint returns a lot of data, most of it is not relevant as you are looking for info about a specific version only.
If you know you want to get the package info only on the latest version, you can use the /latest endpoint which is much more efficient. The only catch is that you can't use it with scoped packages.
In addition, there is no endpoint to get info about a specific version.

My guess is that the first version of `npm-base` used the extended endpoint only. Then someone discovered the latest endpoint and that is why the comment is there. The only thing they missed is handling tags for non scoped packages.

Indeed, non of the code that uses `fetchPackageData` besides the node service, use tags and that is why the bug didn't really effect them.
The `npm-version` service needs refactoring to use the `fetchPackageData` so it will be more efficient and without my fix, this refactoring will introduce the bug to that service as well.

To sum up, I believe my fix is legit and will not harm any other services. Also, we need to open up a feature request to refactor the npm service to utilize the fixed version of `fetchPackageData` and also to add better unit tests to then `fetchPackageData` funciton.

regevbr commented:
Please note I also change the label to be `node-lts` as it will look better when using with tagged versions, e.g `node-lts@next` vs `node lts@next`

calebcartwright commented:
> To sum up, I believe my fix is legit and will not harm any other services.

So the other npm badges are still using the `/latest` endpoint right? if so then i'm good, just want to ensure this change wasn't going to cause the all of those other npm badges to use the other endpoint and larger response size

regevbr commented:
They will not be affected

regevbr commented:
@calebcartwright sorry for the short answer previously.
The other packages will only be affected as they are not passing tag at all those the condition I added will not be invoked. They will use /latest when the package being checked is not scoped, otherwise they will use the extended endpoint, but that is the current behavior in any case.

Even if in the future they will support tag, they will use the extended endpoint, but that is what they should use, otherwise they will get false data.

calebcartwright commented:
> @calebcartwright sorry for the short answer previously.

No worries! You answered my question anyway. Will gives this a final lookover later today

calebcartwright commented on services/node/node-current.tester.js:
```diff
@@ -13,98 +13,164 @@ function expectSemverRange(message) {
 }
 
-t.create('engines satisfies current node version')
+t.create('gets the node version of passport - live')
```
you can drop the ` - live` suffix from the test name. our test runner actually detects whether tests are live or mocked and annotates them accordingly

regevbr replied:
will do

calebcartwright commented on services/node/node-current.tester.js:
```diff
@@ -13,98 +13,164 @@ function expectSemverRange(message) {
     expectSemverRange(json.message)
   })
 
 t.create('engines satisfies current node version with custom registry')
```
We really only need the mocked tests to validate logic within `node-version-color`, and we don't want to have any extra mocked tests that are duplicating scenarios which are explicitly validated by a live test.

As such let's remove the test at the bottom of the file that's mocking an invalid package, and I'm also wondering if this one can be deleted too given the live test above

regevbr replied:
I agree will remove them

calebcartwright commented on services/node/node-lts.service.js:
```diff
@@ -9,7 +9,7 @@ module.exports = class NodeLtsVersion extends NodeVersionBase {
 
   static get defaultBadgeData() {
-    return { label: 'node lts' }
+    return { label: 'node-lts' }
```
:+1:

regevbr commented:
@calebcartwright thanks for your review, I have revised the code accordingly

calebcartwright commented on services/node/node-current.tester.js:
```diff
@@ -0,0 +1,157 @@
+  .intercept(mockCurrentSha(13))
+  .expectBadge({ label: 'node', message: `12`, color: `yellow` })
+
+t.create("gets the tagged release's node version version of ionic")
```
I'm a little surprised that prettier didn't flag the double quotes here (and below) :thinking:

calebcartwright commented on services/node/node-current.tester.js:
```diff
@@ -0,0 +1,157 @@
+  .intercept(mockCurrentSha(13))
+  .expectBadge({ label: 'node', message: `>=0.4.0`, color: `brightgreen` })
+
+t.create('engines not satisfies current node version - scoped')
```
```suggestion
t.create('engines does not satisfy current node version - scoped')
```

calebcartwright commented on services/node/node-current.tester.js:
```diff
@@ -0,0 +1,157 @@
+  .intercept(mockCurrentSha(13))
+  .expectBadge({ label: 'node', message: `>=0.4.0`, color: `brightgreen` })
+
+t.create('engines not satisfies current node version')
```
```suggestion
t.create('engines does not satisfy current node version')
```

calebcartwright commented on services/node/node-current.tester.js:
```diff
@@ -0,0 +1,157 @@
+    color: `brightgreen`,
+  })
+
+t.create('engines not satisfies current node version - tagged')
```
```suggestion
t.create('engines does not satisfy current node version - tagged')
```

calebcartwright commented on services/node/node-current.tester.js:
```diff
@@ -0,0 +1,157 @@
+    color: `brightgreen`,
+  })
+
+t.create('engines not satisfies current node version - scoped and tagged')
```
```suggestion
t.create('engines does not satisfy current node version - scoped and tagged')
```

calebcartwright commented on services/node/node-lts.tester.js:
```diff
@@ -0,0 +1,217 @@
+  .intercept(mockVersionsSha())
+  .expectBadge({ label: 'node-lts', message: `10 - 12`, color: `brightgreen` })
+
+t.create('engines not satisfies all lts node versions')
```
```suggestion
t.create('engines does not satisfy all lts node versions')
```

calebcartwright commented on services/node/node-lts.tester.js:
```diff
@@ -0,0 +1,217 @@
+  .intercept(mockVersionsSha())
+  .expectBadge({ label: 'node-lts', message: `10 - 12`, color: `brightgreen` })
+
+t.create('engines not satisfies all lts node versions - scoped')
```
```suggestion
t.create('engines does not satisfy all lts node versions - scoped')
```

calebcartwright commented on services/node/node-lts.tester.js:
```diff
@@ -0,0 +1,217 @@
+    color: `brightgreen`,
+  })
+
+t.create('engines not satisfies all lts node versions - tagged')
```
```suggestion
t.create('engines does not satisfy all lts node versions - tagged')
```

calebcartwright commented on services/node/node-lts.tester.js:
```diff
@@ -0,0 +1,217 @@
+    color: `brightgreen`,
+  })
+
+t.create('engines not satisfies all lts node versions - scoped and tagged')
```
```suggestion
t.create('engines does not satisfy all lts node versions - scoped and tagged')
```

regevbr commented:
Thank @calebcartwright for the patience and thorough review!

regevbr commented:
We completely forgot to add text documentations to the badges. I will create a new pr
